modules/gateway/synchronize.go

```diff
@@ -34,26 +31,93 @@ func (g *Gateway) Synchronize() (err error) {
 // most recent block seen by both peers. From this starting height, it
 // transmits blocks sequentially. Multiple such transmissions may be required
 // to fully synchronize.
-func (g *Gateway) synchronize(peer network.Address) {
-	var newBlocks []consensus.Block
-	err := peer.RPC("SendBlocks", g.blockHistory(), &newBlocks)
-	if err != nil && err.Error() != moreBlocksErr.Error() {
-		// TODO: try a different peer?
+func (g *Gateway) synchronize(peer modules.NetAddress) {
+	for {
+		var newBlocks []consensus.Block
+		newBlocks, moreAvailable, err := g.requestBlocks(peer)
+		if err != nil {
+			// TODO: try a different peer?
+			return
+		}
+		for _, block := range newBlocks {
+			acceptErr := g.state.AcceptBlock(block)
+			if acceptErr != nil {
+				// TODO: If the error is a FutureTimestampErr, need to wait before trying the
+				// block again.
+			}
+		}
+
+		// loop until there are no more blocks available
+		if !moreAvailable {
+			break
+		}
+	}
+}
+
+// sendBlocks returns a sequential set of blocks based on the 32 input block
+// IDs. The most recent known ID is used as the starting point, and up to
+// 'MaxCatchUpBlocks' from that BlockHeight onwards are returned. It also
+// sends a boolean indicating whether more blocks are available.
+func (g *Gateway) sendBlocks(conn modules.NetConn) (err error) {
+	// read known blocks
+	var knownBlocks [32]consensus.BlockID
+	err = conn.ReadObject(&knownBlocks, 32*crypto.HashSize)
+	if err != nil {
 		return
 	}
-	for _, block := range newBlocks {
-		acceptErr := g.state.AcceptBlock(block)
-		if acceptErr != nil {
-			// TODO: If the error is a FutureTimestampErr, need to wait before trying the
-			// block again.
+
+	// Find the most recent block from knownBlocks that is in our current path.
+	found := false
+	var start consensus.BlockHeight
+	for _, id := range knownBlocks {
+		if height, exists := g.state.HeightOfBlock(id); exists {
+			found = true
+			start = height + 1 // start at child
+			break
 		}
 	}
+	if !found {
+		// The genesis block should be included in knownBlocks - if no matching
+		// blocks are found, the caller is probably on a different blockchain
+		// altogether.
+		return errors.New("no matching block found")
+	}
 
-	if err != nil && err.Error() == moreBlocksErr.Error() {
-		//fmt.Println("getting more blocks")
-		go g.synchronize(peer)
+	// Send blocks, starting with the child of the most recent known block.
+	//
+	// TODO: use BlocksSince instead?
```

Comment:
I think I meant it would be nice to have a function similar to `BlockSince` that would return a slice of blocks. So you could call `s.BlocksInRange(start, start+MaxCatchUpBlocks)` and get back a `[]Block`, instead of doing this silly loop + map lookup business